### ISSUE TITLE
Enhancement: [@typescript-eslint/typedef] Interfaces should be declared in separate files

### ISSUE THREAD

**Issue by luiscla27** [typescript-eslint/typescript-eslint#7746]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://main--typescript-eslint.netlify.app/rules/typedef#options

### Description

Please add a new option called `interfaceLocalDeclaration` with values `true/false`.

This suggestion originates from circular dependencies issues: 

- [Current ESLint `no-cycle` rule is slow](https://www.reddit.com/r/webdev/comments/149hwl9/eslintpluginimport_performance/)
- [Also, the current "best" way to detect them is to use an external tool called madge](https://stackoverflow.com/a/67672027/1657465)
- Currently there's no fast checking for circular dependencies

At least in our team use case, by running [`madge`](https://github.com/pahen/madge), a great amount of circular dependencies were created due to local interfaces being reused in other components. The solution was to simply move those interfaces to another file.

### Fail

```typescript
export interface ButtonProps {
  onClick: () => void;
}

export class Button implements ButtonProps {
  onClick = () => console.log('button!');
}
```

### Pass

```typescript
interface ButtonProps {
  onClick: () => void;
}

export class Button implements ButtonProps {
  onClick = () => console.log('button!');
}
```

### Additional Info

Basically, the rule would prevent to `export` and `implement` an interface in the same file, if the user wants to export it, the solution would be to declare it in another file meant to hold only interfaces.

Instead only having:
```
my-component.component.ts
```
Now we also have a file called:

```
my-component.component.ts
my-component.interfaces.ts // <--- all interfaces were moved there
```

**Comment by bradzacher** [comment]
Thanks for the request!
This is not a common style amongst the TS ecosystem.

Also it doesn't fit into the rule - which solely exists to enforce that type annotations exist.

As such we shall reject this request within our project.

---

Feel free to use our tooling to build either a local rule for your project or publish a 3rd party plugin!

**Comment by luiscla27** [comment]
Actually, the example I included is pretty common... I know it doesn't fit completely there, but I couldn't tell where to suggest the enhancement. I'm not sure if completely new rules are allowed to be submitted. 

Is there any other rule where you think it might fit? I would greatly appreciate your feedback.

Also, please don't discard this as "uncommon" you might be surprised by checking the amount of issues related to this in most typescript projects. The reason is not usually a problem, is that the actual "visible" issue only appears when instances are created. But still, the circular dependency is there (at least this happens in Angular).

**Comment by Josh-Cena** [comment]
@luiscla27 Circular dependency of values is bad because of initialization order. Why is circular type dependency bad? I don't think there's a proof of impact, other than "madge errors on this".

**Comment by luiscla27** [comment]
Because a `class` is also a type. 

And so, when a class is instantiated in another class that implements the same interface, the circular type dependency becomes an issue. The rule is meant to prevent that... 

Of course, that the rule I posted is just a suggestion. Any other rule that might prevent the scenario I mentioned would be okay.

**Comment by Josh-Cena** [comment]
Could you give an example? Does type checking fail in this case?

**Comment by bradzacher** [comment]
There is an existing lint rule to detect cycles `import/no-cycles` - which handle this case and more.
"not exporting" an interface that is implemented would actually be considered a code-smell in a lot of cases - often times the point of declaring an interface is so that you can reference the class shape without referencing the class type (eg keep the class itself as an implementation detail).

Moving the interface to a new file is definitely an option for some usecases - however it's not something that'll suit many codebases. A lot of codebases like to co-locate such things to make maintenance easier (i.e. you only need modify one file when adding a new property instead of two).

TS itself handles TYPE cycles just fine in interfaces and classes as an intentional design decision - so I'm not sure I understand the issue. As Josh said - runtime definitely cycles are bad - but types aren't a problem. ESP if you're using type-only imports TS will elide the imports and thus your code is guaranteed runtime cycle-less.

Unless this pattern specifically causes issues within TS (without some examples I don't know of any cases that would) - I don't see a reason we would build a new rule or add a feature to an existing rule within this project. There's just not a broadly applicable usecase here that would warrant this existing within our project.

**Comment by luiscla27** [comment]
Thank you both so much for your quick responses 🙏 

Yes! in fact you both are right :D

This is a runtime issue, what am trying to explain is that allowing the user to export interfaces that are implemented in the same class of where its being declared, allows the user to code circular runtime dependencies. As I said. I'm not sure if this should be reported/suggested to be handle into an eslint rule here.

@bradzacher, in our codebase we do like to code in a single file, but it caused this issue... I'll try to code an example, but if eslint is not meant to prevent runtime issues then I would just humbly asks you to guide me to which project I should report this. In the links I provided other users suggest that lint-projects are the best option, so that's why I came here.

**Comment by bradzacher** [comment]
I think you're misunderstanding what I say when I say that this doesn't belong in this project.

We're a very general-purpose project lead by volunteers so we don't want to expand our scope to more niche usecases such as this.

I'm not saying that this is something that eslint can't or shouldn't do. I'm saying the the check you want is not a check that most codebases will care about - it's enforcing a code style that most codebases wouldn't necessarily be interested in.

It's definitely a check that eslint can do - you'll just need to build it yourself! As I mentioned before you can use our tooling to build the rule and keep it internal or publish it as a 3rd party plugin.

**Comment by luiscla27** [comment]
Is not a niche use case! Don't stick to that please... Most code bases don't even realize the issue until they're forced to a fix it when instances are created. In our team case, it happened about 4 years old of code base constantly growing.

Maybe the niche would be "big projects". As the circular paths madge errors shown are a minimum of 9 levels long. (what do you think?) 

The solution for the issue doesn't have to be the rule I propose... That's the simplest thing I could come up to solve it. Do you can think of a different suggestion to solve it? Anything is welcomed.

(thank you for clarifying eslint scope).
